inject/src/main/java/io/micronaut/context/Provider.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context;
+
+/**
+ * A provider contract.
+ *
+ * @param <T> The type
+ * @author James Kleeh
+ * @since 2.3.0
+ */
+public interface Provider<T> {
```

Comment:
Maybe name it `BeanProvider` to more easily distinguish it from the other interfaces of the same name